deployment-scanner/src/main/java/org/jboss/as/server/deployment/scanner/FileSystemDeploymentScanHandler.java

```diff
@@ -0,0 +1,89 @@
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
+ * MA 02110-1301  USA
+ */
+package org.jboss.as.server.deployment.scanner;
+
+import static org.jboss.as.controller.descriptions.ModelDescriptionConstants.OP_ADDR;
+
+import java.util.List;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+import org.jboss.as.controller.OperationContext;
+import org.jboss.as.controller.OperationDefinition;
+import org.jboss.as.controller.OperationFailedException;
+import org.jboss.as.controller.OperationStepHandler;
+import org.jboss.as.controller.PathAddress;
+import org.jboss.as.controller.SimpleOperationDefinitionBuilder;
+import org.jboss.as.server.deployment.scanner.logging.DeploymentScannerLogger;
+import org.jboss.dmr.ModelNode;
+import org.jboss.msc.service.ServiceController;
+
+/**
+ * Operation handler to run a single scan by a DeploymentScanner.
+ * @author <a href="mailto:ehugonne@redhat.com">Emmanuel Hugonnet</a> (c) 2014 Red Hat, inc.
+ */
+public class FileSystemDeploymentScanHandler implements OperationStepHandler {
+
+    public static final String OPERATION_NAME = "run-scan";
+    public static final OperationDefinition DEFINITION
+            = new SimpleOperationDefinitionBuilder(OPERATION_NAME, DeploymentScannerExtension.getResourceDescriptionResolver("deployment.scanner")).build();
+
+    public static final FileSystemDeploymentScanHandler INSTANCE = new FileSystemDeploymentScanHandler();
+
+    private FileSystemDeploymentScanHandler() {
+    }
+
+    @Override
+    public void execute(OperationContext context, ModelNode operation) throws OperationFailedException {
+        context.addStep(new OperationStepHandler() {
+            @Override
+            public void execute(OperationContext context, ModelNode operation) throws OperationFailedException {
+                final FileSystemDeploymentService scanner = getExistingScanner(context, operation);
+                if (scanner != null) {
+                    final Future<List<?>> tasks = scanner.singleScan();
+                    context.completeStep(new OperationContext.RollbackHandler() {
+
+                        @Override
+                        public void handleRollback(OperationContext context, ModelNode operation) {
+                            try {
+                                scanner.rollBack(tasks.get(), context.getResult());
```

Comment:
I don't get why this doesn't deadlock, although your test case implies it doesn't. The tasks.get() call should block until ManualScanCallable completes, and it won't complete because this op is holding the exclusive lock.

Oh, I see why the test passes and doesn't deadlock. I'll comment there.